pkgs/misc/crust/default.nix

```
@@ -0,0 +1,154 @@
+
+    defconfig = if defconfig == null then
+      throw "crust config not defined!"
+      else defconfig;
+
+    version = if src == null then
+      defaultVersion
+      else version;
+
+    src = if src == null then
+      defaultSrc
+      else src;
+
+    # Only for global crust patches, per-device paches should be included as extras to avoid adding noise which would complicate debugging
+    patches = [] ++ extraPatches;
+
+    postPatch = "";
+
+    nativeBuildInputs = [
+      flex
+      bison
+    ];
+
+    depsBuildBuild = [ buildPackages.stdenv.cc ];
+
+    makeFlags = [
+      # This is double-quotted to prevent string splitting (SC2086)
+      "CROSS_COMPILE=\"${stdenv.cc.targetPrefix}\""
+    ] ++ extraMakeFlags;
+
+    passAsFile = [ "extraConfig" ];
+
+    ###! Phases
+    configurePhase = ''
+      runHook preConfigure
+
+      echo "Configuring for ${defconfig}"
+        make ${defconfig}
+
+      # The variable 'extraConfigPath' is often parsed as blank resulting in a scenario where `cat` is expecting interactive input followed by line-break to continue, this is to reduce the risk of that happening
+      [ -z "$extraConfigPath" ] || {
+        cat $extraConfigPath >> .config
+      }
+
+      runHook postConfigure
+    '';
+
+    installPhase = ''
+      runHook preInstall
+
+      echo "Creating directory in ${installDir}"
+      mkdir -p "${installDir}"
+
+      # Process 'filesToInstall' from declarations
+      for file in ${lib.concatStringsSep " " filesToInstall}; do
+        cp -v "$file" "${installDir}"
+      done
+
+      runHook postInstall
+    '';
+
+    ###! Hardening
+    ###! * 'bindnow' is ignored by the package
+    ###! * 'pie' is already applied by default to musl and disabled for glibc where in our scenario it seems to be unsupported
+    ###! * 'relro' is ignored by the package, likely due to it breaking dynamic objects. Issue https://github.com/crust-firmware/crust/issues/216 has been submitted to verify
+    ###! Refer to https://nixos.org/manual/nixpkgs/stable/#sec-hardening-in-nixpkgs for details
+    hardeningDisable = [
+      "bindnow"
+      "relro"
+      "pie"
+    ];
+
+    # Default Metadata
+    meta = with lib; {
+      homepage = "https://github.com/crust-firmware/crust";
+      # Try to adjust the description per package so that it's not all with generic wording
+      description = "SCP (power management) firmware for sunxi SoCs";
+      license = licenses.bsd3;
+    } // extraMeta;
+  } // removeAttrs args [ "extraMeta" ]);
+
+in {
+  inherit buildCrust;
+
+  # Example configuration:
+  ## crustExample = buildCrust rec {
+  ##   defconfig = "example_defconfig";
+  ##   extraMeta = {
+  ##     description = "SCP (power management) firmware for Example";
+  ##     platforms = [ "or1k-none" ];
+  ##     maintainers = with lib.maintainers; [ your-usename-here ];
+  ##   };
+  ##   filesToInstall = [ "files/to/export" ];
+  ## };
+
+  crustTeresA64 = buildCrust rec {
```

```suggestion
  crustTeresA64 = buildCrust {
```

Use `rec` only when needed.

---

addressed in upcoming